### ISSUE TITLE
HTTP request callback function inside reporter not called

### ISSUE THREAD

**Issue by vdespa** [postmanlabs/newman#2148]
1. Newman Version (can be found via `newman -v`): 4.5.5
 2. OS details (type, version, and architecture): macOS
 3. Are you using Newman as a library, or via the CLI? CLI

I am trying to use the node request library from within a Newman reporter. Below is a sample code:

```
    newman.on('beforeDone', () => {
      console.log('before');

      var request = require('request');
      request('http://httpbin.org/get', function (error, response, body) {
        console.log('error:', error); // Print the error if one occurred
        console.log('statusCode:', response && response.statusCode); // Print the response status code if a response was received
        console.log('body:', body); // Print the HTML for the Google homepage.
      });

      console.log('after');
    });
```
Maybe I am missing something, but the console.log statements inside the callback function are not called. Only `before` and `after` is printed.

Do you have any idea why?

**Comment by codenirvana** [comment]
@vdespa is execution getting terminated before `request` callback?

I can't think of any other reason for this issue. You can see we are doing a similar request in `beforeDone` callback in this reporter: https://github.com/postmanlabs/newman-reporter-remote/blob/develop/lib/index.js#L431

**Comment by vdespa** [comment]
Yes, it looks that Newman is not waiting for the callback and terminating the execution. When enabling HTTP debugging I can see that the request is being sent.

I had looked into other reporters that use HTTP and they share the same issue. 

Thank you for the link, I will look at it in the next days.

**Comment by codenirvana** [comment]
@vdespa can you share the link to the reporter (or a sample reporter) with this behavior, will have a look.

**Comment by vdespa** [comment]
Thank you for looking into this.

Here is the index.js file:

```
function MyNewmanReporter (newman, reporterOptions, collectionRunOptions) {
 
  
    newman.on('beforeDone', () => {
      console.log('before');

      var request = require('request');
      request('http://httpbin.org/get', function (error, response, body) {
        console.log('error:', error); // Print the error if one occurred
        console.log('statusCode:', response && response.statusCode); // Print the response status code if a response was received
        console.log('body:', body); // Print the HTML for the Google homepage.
      });

      console.log('after');
    });

  }
  module.exports = MyRailSestraNewmanReporter
```

I have noticed the same behavior in this reporter as well: https://www.npmjs.com/package/newman-reporter-testrail-extra

**Comment by codenirvana** [comment]
With this I got the following logged before ending the execution: (which I feel is the expected output)

```
before
after
error: null
statusCode: 200
body: {
  "args": {},
  "headers": {
    "Host": "httpbin.org"
  },
  "origin": "127.0.0.1",
  "url": "https://httpbin.org/get"
}
```

Using Newman CLI on Node v10 & v12.

**Comment by vdespa** [comment]
I had a very strange behavior: I enabled the CLI reporter and it worked. Now it works. I don't get it :)

Thank you for your help. I will close this at this moment as I can no longer replicate the problem.

**Comment by vdespa** [comment]
@codenirvana 
I have managed to recreate this problem in a more stable environment:

https://gitlab.com/user-questions/newman-issues-2148

Here is an example console output and as you can notice, the callback is not printed.
https://gitlab.com/user-questions/newman-issues-2148/-/jobs/323473763

Am I missing something?

**Comment by codenirvana** [comment]
@vdespa I can confirm this happens when there's a failure (like assertion failure) in the Newman execution because the process gets terminated with exit code 1.

For now, you can use the `-x`, `--suppress-exit-code` option to avoid this behavior.

I feel for async tasks like this, `beforeDone` should accept a callback function e.g:
```js
newman.on('beforeDone', (done) => {
  request('http://httpbin.org/get', (error, response, body) => {
    done();
  });
});
```

**Comment by vdespa** [comment]
Is there any way to do this without (-x).

It would be great if you could provide a "best practice" example for doing this. I see many 3rd party reporters struggling with this.

**Comment by zeburek** [comment]
@codenirvana Is there any update on this issue?
I'm also faced issue described in https://github.com/postmanlabs/newman/issues/2148#issuecomment-543075798

I'm writing new reporter for quite new TMS and this looks like a very unexpected behaviour.
I also have lots of async request, which are correctly served when all tests are passed and not - when something is failed.

**Comment by arirarira** [comment]
hello, i have this code :
